core/src/main/java/jeeves/server/context/ServiceContext.java

```diff
@@ -115,10 +160,168 @@ public static ServiceContext get() {
 
     /**
      * Called to set the Service context for this thread and inherited threads.
+     *
+     * If you call this method you are responsible for thread context management and {@link #clearAsThreadLocal()}.
+     * <pre>
+     * try {
+     *     context.setAsThreadLocal();
+     * }
+     * finally {
+     *     context.clearAsThreadLocal();
+     * }
+     * </pre>
      */
     public void setAsThreadLocal() {
+        ServiceContext check = THREAD_LOCAL_INSTANCE.get();
+
+        if( POLICY == ThreadLocalPolicy.DIRECT || check == null){
+            // step one set thread local
+            THREAD_LOCAL_INSTANCE.set(this);
+            // step two ensure ApplicationContextHolder thread local kept in sync
+            ApplicationContextHolder.set(this.getApplicationContext());
+            // step three details on allocation
+            allocation = new Throwable("ServiceContext allocated to thread");
```

Comment:
This is pretty ugly coding, keeping a throwable in order to report back who called setAsThreadLocal() in case of future conflict.